lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp

```diff
@@ -382,227 +383,289 @@ const CompilerType &SwiftLanguageRuntimeImpl::GetBoxMetadataType() {
+          address.getAddressData())) {
+    addr = *resolved_addr;
+  }
 
-    LLDB_LOGV(log, "[MemoryReader] asked to read {0} bytes at address {1:x}",
-              size, address.getAddressData());
+  if (size > m_max_read_amount) {
+    LLDB_LOGV(log, "[MemoryReader] memory read exceeds maximum allowed size");
+    return false;
+  }
+  Target &target(m_process.GetTarget());
+  Status error;
+  if (size > target.ReadMemory(addr, dest, size, error, true)) {
+    LLDB_LOGV(log,
+              "[MemoryReader] memory read returned fewer bytes than asked for");
+    return false;
+  }
+  if (error.Fail()) {
+    LLDB_LOGV(log, "[MemoryReader] memory read returned error: {0}",
+              error.AsCString());
+    return false;
+  }
 
-    if (size > m_max_read_amount) {
-      LLDB_LOGV(log, "[MemoryReader] memory read exceeds maximum allowed size");
-      return false;
+  auto format_data = [](auto dest, auto size) {
+    StreamString stream;
+    for (uint64_t i = 0; i < size; i++) {
+      stream.PutHex8(dest[i]);
+      stream.PutChar(' ');
     }
+    return stream.GetData();
+  };
+  LLDB_LOGV(log, "[MemoryReader] memory read returned data: {0}",
+            format_data(dest, size));
 
-    Target &target(m_process.GetTarget());
-    Address addr(address.getAddressData());
-    Status error;
-    if (size > target.ReadMemory(addr, dest, size, error, true)) {
-      LLDB_LOGV(log, "[MemoryReader] memory read returned fewer bytes than asked for");
-      return false;
-    }
-    if (error.Fail()) {
-      LLDB_LOGV(log, "[MemoryReader] memory read returned error: {0}",
-                error.AsCString());
-      return false;
-    }
+  return true;
+}
+
+bool LLDBMemoryReader::readString(swift::remote::RemoteAddress address,
+                                  std::string &dest) {
+  Log *log(GetLogIfAllCategoriesSet(LIBLLDB_LOG_TYPES));
 
-    auto format_data = [](auto dest, auto size) {
+  LLDB_LOGV(log, "[MemoryReader] asked to read string data at address {0x}",
+            address.getAddressData());
+
+  Address addr = Address(address.getAddressData());
+  if (auto resolved_addr = tryToConvertFromMappedAddressIntoRealAddress(
+          address.getAddressData()))
+    addr = *resolved_addr;
+
+  Target &target(m_process.GetTarget());
+  Status error;
+  target.ReadCStringFromMemory(addr, dest, error);
+  if (error.Success()) {
+    auto format_string = [](const std::string &dest) {
       StreamString stream;
-      for (uint64_t i = 0; i < size; i++) {
-        stream.PutHex8(dest[i]);
-        stream.PutChar(' ');
+      for (auto c : dest) {
+        if (c >= 32 && c <= 127) {
+          stream << c;
+        } else {
+          stream << "\\0";
+          stream.PutHex8(c);
+        }
       }
       return stream.GetData();
     };
-    LLDB_LOGV(log, "[MemoryReader] memory read returned data: {0}",
-              format_data(dest, size));
-
+    LLDB_LOGV(log, "[MemoryReader] memory read returned data: \"{0}\"",
```

Comment:
Would you mind factoring out the NFC improvements into a separate, obviously correct, patch that we can land first?